File: web/src/topologies.ts
// Topologies — the browser-side mirror of the `/agent/topologies` routes
// (work plan 13 T2/T3; engine: go/topology, go/httpapi/topologies.go).
//
// Pure: no React, no window, no fetch. The hook lives in useTopologies.ts and
// the flow component in components/TopologyOnboarding.tsx.
//
// Validation policy: mirror the engine, do not out-legislate it. The server's
// ResolveAnswers enforces exactly three things a form can trip over — a
// required question left unanswered (with no default), an answer of the wrong
// type, and a choice outside the question's Choices — so this module enforces
// exactly those. The form's controls make wrong types unrepresentable, which
// leaves "required and blank" and "not one of the choices" as the two errors
// worth naming before the server does.

import { coerceWorker, type Worker } from './workers.js'
import { coerceSubscription, type Subscription } from './events.js'
import { coerceSchedule, type Schedule } from './schedules.js'
import { coerceConfigEvent, type ConfigEvent } from './configLog.js'

/** Endpoint paths for the topology routes. Overridable per host, like
 *  WORKER_ENDPOINTS. */
export const TOPOLOGY_ENDPOINTS = {
  list: '/agent/topologies',
  preview: '/agent/topologies/preview',
  apply: '/agent/topologies/apply',
}

/** The closed vocabulary of answer shapes (go/topology QuestionType). */
export type TopologyQuestionType = 'string' | 'bool' | 'choice'

/** One thing a topology asks before it can render. Field names are the wire
 *  names. `default` is null when the question has none — the wire omits it. */
export interface TopologyQuestion {
  id: string
  prompt: string
  type: TopologyQuestionType
  /** Populated for choice questions only. */
  choices: string[]
  default: string | boolean | null
  required: boolean
}

/** One catalogue entry: a named, versioned org chart (D1: defined in code).
 *  Identity is name@version, same naming posture as images and skills. */
export interface Topology {
  name: string
  version: string
  description: string
  questions: TopologyQuestion[]
}

/** The topology's name@version identity string, e.g. "solo@v1". */
export function topologyRef(t: Pick<Topology, 'name' | 'version'>): string {
  return `${t.name}@${t.version}`
}

/** Question ID → the answer the form holds. Strings for string/choice
 *  questions ('' = unanswered), booleans for bool questions. */
export type TopologyAnswers = Record<string, string | boolean>

/** One would-be subscription in the preview diff. */
export interface TopologyRouteSummary {
  event_type: string
  worker: string
}

/** One would-be schedule in the preview diff. */
export interface TopologyCronSummary {
  cron: string
  worker: string
  input: string
}

/** The preview against the project's CURRENT config: what would be created,
 *  and which worker names are already taken (workers are the one name-keyed
 *  kind, hence the one that can collide). */
export interface TopologyDiff {
  new_workers: string[]
  colliding_workers: string[]
  new_subscriptions: TopologyRouteSummary[]
  new_schedules: TopologyCronSummary[]
  /** Project-settings fields the bundle's patch would set. */
  settings_fields: string[]
  memory_seeds: number
}

/** One rendered org chart: rows of the existing config types, project-agnostic
 *  (apply stamps project/IDs/timestamps). Mirrors go/topology.Bundle. */
export interface TopologyBundle {
  workers: Worker[]
  subscriptions: Subscription[]
  schedules: Schedule[]
  settings_patch: Record<string, unknown> | null
  memory_seeds: unknown[]
  preconditions: { images: string[]; skills: string[] }
}

/** The whole preview response. `applicable` is the one-word verdict: no
 *  collisions and no missing preconditions — apply MUST be blocked when false. */
export interface TopologyPreview {
  topology: Topology
  bundle: TopologyBundle
  diff: TopologyDiff
  missing_images: string[]
  missing_skills: string[]
  applicable: boolean
}

/** Everything the apply created, read back (§9), plus the bracketing
 *  `topology_apply` config event — the receipt. */
export interface TopologyApplyResult {
  workers: Worker[]
  subscriptions: Subscription[]
  schedules: Schedule[]
  settings: Record<string, unknown> | null
  memories: unknown[]
  event: ConfigEvent
}

// ---------------------------------------------------------------------------
// Coercion — fill anything the server omitted so components never branch on
// undefined, the same posture as coerceWorker.
// ---------------------------------------------------------------------------

const str = (v: unknown, fallback = ''): string => (typeof v === 'string' ? v : fallback)
const strings = (v: unknown): string[] => (Array.isArray(v) ? v.map(String) : [])
const record = (v: unknown): Record<string, unknown> | null =>
  v && typeof v === 'object' && !Array.isArray(v) ? (v as Record<string, unknown>) : null

export function coerceTopologyQuestion(raw: unknown): TopologyQuestion {
  const r = record(raw) ?? {}
  const type = r.type === 'bool' || r.type === 'choice' ? r.type : 'string'
  return {
    id: str(r.id),
    prompt: str(r.prompt),
    type,
    choices: strings(r.choices),
    default: typeof r.default === 'string' || typeof r.default === 'boolean' ? r.default : null,
    required: r.required === true,
  }
}

export function coerceTopology(raw: unknown): Topology {
  const r = record(raw) ?? {}
  return {
    name: str(r.name),
    version: str(r.version),
    description: str(r.description),
    questions: Array.isArray(r.questions) ? r.questions.map(coerceTopologyQuestion) : [],
  }
}

export function coerceTopologyDiff(raw: unknown): TopologyDiff {
  const r = record(raw) ?? {}
  return {
    new_workers: strings(r.new_workers),
    colliding_workers: strings(r.colliding_workers),
    new_subscriptions: Array.isArray(r.new_subscriptions)
      ? r.new_subscriptions.map((s) => {
          const e = record(s) ?? {}
          return { event_type: str(e.event_type), worker: str(e.worker) }
        })
      : [],
    new_schedules: Array.isArray(r.new_schedules)
      ? r.new_schedules.map((s) => {
          const e = record(s) ?? {}
          return { cron: str(e.cron), worker: str(e.worker), input: str(e.input) }
        })
      : [],
    settings_fields: strings(r.settings_fields),
    memory_seeds: typeof r.memory_seeds === 'number' ? r.memory_seeds : 0,
  }
}

export function coerceTopologyBundle(raw: unknown): TopologyBundle {
  const r = record(raw) ?? {}
  const pre = record(r.preconditions) ?? {}
  return {
    workers: Array.isArray(r.workers) ? r.workers.map((w) => coerceWorker(w)) : [],
    subscriptions: Array.isArray(r.subscriptions) ? r.subscriptions.map(coerceSubscription) : [],
    schedules: Array.isArray(r.schedules) ? r.schedules.map((s) => coerceSchedule(s)) : [],
    settings_patch: record(r.settings_patch),
    memory_seeds: Array.isArray(r.memory_seeds) ? r.memory_seeds : [],
    preconditions: { images: strings(pre.images), skills: strings(pre.skills) },
  }
}

export function coerceTopologyPreview(raw: unknown): TopologyPreview {
  const r = record(raw) ?? {}
  return {
    topology: coerceTopology(r.topology),
    bundle: coerceTopologyBundle(r.bundle),
    diff: coerceTopologyDiff(r.diff),
    missing_images: strings(r.missing_images),
    missing_skills: strings(r.missing_skills),
    // Absent/garbled reads as NOT applicable — the failure mode must block
    // apply, never wave it through.
    applicable: r.applicable === true,
  }
}

export function coerceTopologyApplyResult(raw: unknown): TopologyApplyResult {
  const r = record(raw) ?? {}
  return {
    workers: Array.isArray(r.workers) ? r.workers.map((w) => coerceWorker(w)) : [],
    subscriptions: Array.isArray(r.subscriptions) ? r.subscriptions.map(coerceSubscription) : [],
    schedules: Array.isArray(r.schedules) ? r.schedules.map((s) => coerceSchedule(s)) : [],
    settings: record(r.settings),
    memories: Array.isArray(r.memories) ? r.memories : [],
    event: coerceConfigEvent(r.event),
  }
}

// ---------------------------------------------------------------------------
// The answer form
// ---------------------------------------------------------------------------

/**
 * The form state a topology's questions start from: defaults pre-filled, bool
 * questions seeded false (a switch always shows a definite state), string and
 * choice questions '' (which means "unanswered" — see topologyAnswersBody).
 */
export function initialTopologyAnswers(questions: TopologyQuestion[]): TopologyAnswers {
  const answers: TopologyAnswers = {}
  for (const q of questions) {
    if (q.default !== null) answers[q.id] = q.default
    else if (q.type === 'bool') answers[q.id] = false
    else answers[q.id] = ''
  }
  return answers
}

/** Question ID → human-readable problem. Empty when the form can be sent. */
export type TopologyAnswerErrors = Record<string, string>

/**
 * The two form-level failures the server would refuse: a required string or
 * choice question left blank, and a choice answer outside the choices (which a
 * select cannot produce, but a host driving answers programmatically can).
 * Bool questions cannot be invalid: false is an answer, not an absence.
 */
export function validateTopologyAnswers(
  questions: TopologyQuestion[],
  answers: TopologyAnswers,
): TopologyAnswerErrors {
  const errors: TopologyAnswerErrors = {}
  for (const q of questions) {
    const v = answers[q.id]
    if (q.type === 'bool') continue
    const text = typeof v === 'string' ? v.trim() : ''
    if (text === '') {
      if (q.required) errors[q.id] = 'an answer is required'
      continue
    }
    if (q.type === 'choice' && !q.choices.includes(text)) {
      errors[q.id] = `must be one of: ${q.choices.join(', ')}`
    }
  }
  return errors
}

/**
 * The wire `answers` object for preview/apply: bool answers verbatim, string
 * and choice answers trimmed — and omitted when blank, because the server
 * distinguishes "unanswered" (default applies, optional passes) from "answered
 * with the empty string" (an answer the renderer must swallow).
 */
export function topologyAnswersBody(
  questions: TopologyQuestion[],
  answers: TopologyAnswers,
): Record<string, string | boolean> {
  const body: Record<string, string | boolean> = {}
  for (const q of questions) {
    const v = answers[q.id]
    if (q.type === 'bool') {
      if (typeof v === 'boolean') body[q.id] = v
      continue
    }
    const text = typeof v === 'string' ? v.trim() : ''
    if (text !== '') body[q.id] = text
  }
  return body
}

File: go/httpapi/topologies.go
package httpapi

// topologies.go — T2's HTTP surface over the topology library (go/topology)
// and the transactional apply (agentdb.ApplyTopology). Three routes, all on
// the JWT path, project always from the token's Customer claim (P5):
//
//	GET  /agent/topologies          — the built-in catalogue (D1)
//	POST /agent/topologies/preview  — render + diff + preconditions; writes NOTHING
//	POST /agent/topologies/apply    — the same computation, then the atomic apply
//
// Preview and apply share one computation (previewTopology), so what apply
// refuses is exactly what preview flagged. The handler-side check is advisory
// — ApplyTopology re-checks inside its transaction and its sentinels map to
// the same 409 — which closes the race between a preview and a concurrent
// mutation without ever trusting the client.

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"

	"github.com/binocarlos/badcode-agent-orange/agentdb"
	"github.com/binocarlos/badcode-agent-orange/topology"
)

// TopologyStore is the store seam the topology routes need: the reads that
// power the preview diff, and the ONE write — the atomic apply, which routes
// every row through the existing config-logged mutations. *agentdb.Store
// implements it; New() fills it from AgentDB, and without one the routes
// answer 501 like every other product-layer seam.
type TopologyStore interface {
	ListWorkers(ctx context.Context, project string) ([]*agentdb.Worker, error)
	GetProjectSettings(ctx context.Context, project string) (*agentdb.ProjectSettings, error)
	ResolveCustomImage(ctx context.Context, project, ref string) (*agentdb.CustomImage, error)
	GetProjectSkill(ctx context.Context, project, name string) (*agentdb.Skill, error)
	ApplyTopology(ctx context.Context, app agentdb.TopologyApplication, cw agentdb.ConfigWrite) (*agentdb.TopologyApplyResult, error)
}

// topologies returns the configured store, or writes 501 and returns nil
// (mirrors the workers/schedules contract).
func (h *Handlers) topologies(w http.ResponseWriter) TopologyStore {
	if h.cfg.Topologies == nil {
		http.Error(w, "topology store not configured", http.StatusNotImplemented)
		return nil
	}
	return h.cfg.Topologies
}

// topologyBody is the preview/apply payload: which topology, and the answers.
type topologyBody struct {
	Name    string         `json:"name"`
	Version string         `json:"version"`
	Answers map[string]any `json:"answers"`
	// Rationale is the operator's why, threaded like every other human edit
	// (E3). Absent is fine: ApplyTopology defaults it to "seeded from
	// name@version", which is the honest reason for a seeded row anyway.
	Rationale string `json:"rationale,omitempty"`
}

// topologyRouteSummary names one would-be subscription in the diff.
type topologyRouteSummary struct {
	EventType string `json:"event_type"`
	Worker    string `json:"worker"`
}

// topologyCronSummary names one would-be schedule in the diff.
type topologyCronSummary struct {
	Cron   string `json:"cron"`
	Worker string `json:"worker"`
	Input  string `json:"input"`
}

// topologyDiff is the preview against the project's CURRENT config: what would
// be created, and which worker names are already taken. Subscriptions and
// schedules are id-keyed rows, so they are always new; workers are the one
// name-keyed kind and the one that can collide.
type topologyDiff struct {
	NewWorkers       []string               `json:"new_workers"`
	CollidingWorkers []string               `json:"colliding_workers"`
	NewSubscriptions []topologyRouteSummary `json:"new_subscriptions"`
	NewSchedules     []topologyCronSummary  `json:"new_schedules"`
	// SettingsFields are the project-settings fields the bundle's patch would
	// set — computed by the SAME overlay apply performs (TopologySettingsOverlay).
	SettingsFields []string `json:"settings_fields"`
	MemorySeeds    int      `json:"memory_seeds"`
}

// topologyPreview is the whole preview response. Applicable is the one-word
// verdict: no collisions and no missing preconditions.
type topologyPreview struct {
	Topology      *topology.Topology `json:"topology"`
	Bundle        *topology.Bundle   `json:"bundle"`
	Diff          topologyDiff       `json:"diff"`
	MissingImages []string           `json:"missing_images"`
	MissingSkills []string           `json:"missing_skills"`
	Applicable    bool               `json:"applicable"`
}

// ListTopologies returns the built-in catalogue: name, version, description
// and the question list — everything a UI needs to interview the operator.
func (h *Handlers) ListTopologies(w http.ResponseWriter, r *http.Request) {
	if _, ok := h.identify(w, r); !ok {
		return
	}
	writeJSON(w, map[string]any{"topologies": topology.List()})
}

// previewTopology is the shared computation. It performs reads only — the
// property PreviewTopology's tests pin — and every error it returns is already
// mapped to a status code.
func (h *Handlers) previewTopology(ctx context.Context, store TopologyStore, project string, body topologyBody) (*topologyPreview, *topology.Topology, *topology.Bundle, int, error) {
	t, ok := topology.Get(body.Name, body.Version)
	if !ok {
		return nil, nil, nil, http.StatusNotFound, errors.New("topology not found")
	}
	bundle, err := t.Instantiate(topology.Answers(body.Answers))
	if err != nil {
		// Both wraps are the caller's answers being wrong (400 by the topology
		// package's contract); anything else would be a registry bug.
		if errors.Is(err, topology.ErrBadAnswers) || errors.Is(err, topology.ErrRender) {
			return nil, nil, nil, http.StatusBadRequest, err
		}
		return nil, nil, nil, http.StatusInternalServerError, err
	}

	existing, err := store.ListWorkers(ctx, project)
	if err != nil {
		return nil, nil, nil, http.StatusInternalServerError, err
	}
	taken := make(map[string]bool, len(existing))
	for _, w := range existing {
		taken[w.Name] = true
	}

	diff := topologyDiff{
		NewWorkers:       []string{},
		CollidingWorkers: []string{},
		NewSubscriptions: []topologyRouteSummary{},
		NewSchedules:     []topologyCronSummary{},
		SettingsFields:   []string{},
		MemorySeeds:      len(bundle.MemorySeeds),
	}
	for _, bw := range bundle.Workers {
		if taken[bw.Name] {
			diff.CollidingWorkers = append(diff.CollidingWorkers, bw.Name)
		} else {
			diff.NewWorkers = append(diff.NewWorkers, bw.Name)
		}
	}
	for _, sub := range bundle.Subscriptions {
		diff.NewSubscriptions = append(diff.NewSubscriptions, topologyRouteSummary{
			EventType: sub.EventType, Worker: sub.Worker,
		})
	}
	for _, sch := range bundle.Schedules {
		diff.NewSchedules = append(diff.NewSchedules, topologyCronSummary{
			Cron: sch.Cron, Worker: sch.Worker, Input: sch.Input,
		})
	}
	if bundle.SettingsPatch != nil {
		current, err := store.GetProjectSettings(ctx, project)
		if err != nil {
			return nil, nil, nil, http.StatusInternalServerError, err
		}
		_, diff.SettingsFields = agentdb.TopologySettingsOverlay(current, bundle.SettingsPatch)
	}

	missingImages := []string{}
	for _, ref := range bundle.Preconditions.Images {
		if _, err := store.ResolveCustomImage(ctx, project, ref); err != nil {
			if errors.Is(err, agentdb.ErrCustomImageNotFound) || errors.Is(err, agentdb.ErrCustomImageInvalid) {
				missingImages = append(missingImages, ref)
				continue
			}
			return nil, nil, nil, http.StatusInternalServerError, err
		}
	}
	missingSkills := []string{}
	for _, name := range bundle.Preconditions.Skills {
		if _, err := store.GetProjectSkill(ctx, project, name); err != nil {
			if errors.Is(err, agentdb.ErrSkillNotFound) || errors.Is(err, agentdb.ErrSkillInvalid) {
				missingSkills = append(missingSkills, name)
				continue
			}
			return nil, nil, nil, http.StatusInternalServerError, err
		}
	}

	preview := &topologyPreview{
		Topology:      t,
		Bundle:        bundle,
		Diff:          diff,
		MissingImages: missingImages,
		MissingSkills: missingSkills,
		Applicable:    len(diff.CollidingWorkers) == 0 && len(missingImages) == 0 && len(missingSkills) == 0,
	}
	return preview, t, bundle, 0, nil
}

// PreviewTopology renders the bundle for {name, version, answers} and answers
// with the diff and the unmet preconditions. It writes NOTHING — it is the
// look-before-you-leap half of apply.
func (h *Handlers) PreviewTopology(w http.ResponseWriter, r *http.Request) {
	id, ok := h.identify(w, r)
	if !ok {
		return
	}
	store := h.topologies(w)
	if store == nil {
		return
	}
	var body topologyBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	preview, _, _, status, err := h.previewTopology(r.Context(), store, id.Customer, body)
	if err != nil {
		http.Error(w, err.Error(), status)
		return
	}
	writeJSON(w, preview)
}

// ApplyTopologyHandler runs the same computation as preview and, when the
// verdict is applicable, performs the atomic apply. A collision or an unmet
// precondition is a 409 that changes nothing — refused here when the preview
// already shows it, and refused inside the store's transaction when a
// concurrent mutation created it after this check.
func (h *Handlers) ApplyTopologyHandler(w http.ResponseWriter, r *http.Request) {
	id, ok := h.identify(w, r)
	if !ok {
		return
	}
	store := h.topologies(w)
	if store == nil {
		return
	}
	var body topologyBody
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	preview, t, bundle, status, err := h.previewTopology(r.Context(), store, id.Customer, body)
	if err != nil {
		http.Error(w, err.Error(), status)
		return
	}
	if !preview.Applicable {
		refusal := "topology not applicable:"
		for _, name := range preview.Diff.CollidingWorkers {
			refusal += " worker " + name + " already exists;"
		}
		for _, ref := range preview.MissingImages {
			refusal += " image " + ref + " not in the project's catalogue;"
		}
		for _, name := range preview.MissingSkills {
			refusal += " skill " + name + " not in the project's catalogue;"
		}
		http.Error(w, refusal+" nothing was changed", http.StatusConflict)
		return
	}

	// Record the RESOLVED answers (defaults applied) — what was actually
	// rendered, not just what was typed. Instantiate already validated them.
	resolved, err := topology.ResolveAnswers(t.Questions, topology.Answers(body.Answers))
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	app := agentdb.TopologyApplication{
		Project:        id.Customer,
		Topology:       t.Ref(),
		Answers:        agentdb.JSONMap(resolved),
		SettingsPatch:  bundle.SettingsPatch,
		RequiredImages: bundle.Preconditions.Images,
		RequiredSkills: bundle.Preconditions.Skills,
	}
	for i := range bundle.Workers {
		app.Workers = append(app.Workers, &bundle.Workers[i])
	}
	for i := range bundle.Subscriptions {
		app.Subscriptions = append(app.Subscriptions, &bundle.Subscriptions[i])
	}
	for i := range bundle.Schedules {
		app.Schedules = append(app.Schedules, &bundle.Schedules[i])
	}
	for i := range bundle.MemorySeeds {
		app.MemorySeeds = append(app.MemorySeeds, &bundle.MemorySeeds[i])
	}

	result, err := store.ApplyTopology(r.Context(), app, humanEditBecause(body.Rationale))
	if err != nil {
		writeTopologyApplyErr(w, err)
		return
	}
	writeJSON(w, result)
}

// writeTopologyApplyErr maps the store's refusals onto status codes: the two
// not-in-that-state sentinels are 409, validation failures are 400, everything
// else is 500.
func writeTopologyApplyErr(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, agentdb.ErrTopologyNameCollision),
		errors.Is(err, agentdb.ErrTopologyPreconditionUnmet):
		http.Error(w, err.Error(), http.StatusConflict)
	case errors.Is(err, agentdb.ErrWorkerInvalid),
		errors.Is(err, agentdb.ErrScheduleInvalid),
		errors.Is(err, agentdb.ErrInvalidProjectSettings):
		http.Error(w, err.Error(), http.StatusBadRequest)
	default:
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}
